cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -120,45 +139,274 @@ func NewResourceManager(skipID, fullResourceName string) (ResourceManager, error
 // This intentionally only logs errors and returns with the UseDefaultMethodError,
 // in case any errors are hit. This is to avoid clusters filling up with unexpected admission errors.
-func (rm *resourceManager) ReallocateWithFractionalResources(request *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
-	if !isInputOk(request, rm.skipID) {
+func (rm *resourceManager) CreateFractionalResourceResponse(request *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	if !isAllocateRequestOk(request, rm.skipID) {
 		// it is better to leave allocated gpu devices as is and return
 		return nil, &dpapi.UseDefaultMethodError{}
 	}
 
+	klog.V(4).Info("Proposed device ids: ", request.ContainerRequests[0].DevicesIDs)
+
 	podCandidate, err := rm.findAllocationPodCandidate()
-	if _, ok := err.(*retryErr); ok {
+	if errors.Is(err, &retryErr{}) {
 		klog.Warning("retrying POD resolving after sleeping")
-		time.Sleep(retryTimeout)
+		time.Sleep(rm.retryTimeout)
 
 		podCandidate, err = rm.findAllocationPodCandidate()
 	}
 
 	if err != nil {
-		if _, ok := err.(*zeroPendingErr); !ok {
+		if !errors.Is(err, &zeroPendingErr{}) {
 			klog.Error("allocation candidate not found, perhaps the GPU scheduler extender is not called, err:", err)
 		}
 		// it is better to leave allocated gpu devices as is and return
 		return nil, &dpapi.UseDefaultMethodError{}
 	}
 
 	pod := podCandidate.pod
-	cards := containerCards(pod, podCandidate.allocatedContainers)
 
-	return rm.createAllocateResponse(cards)
+	rm.cleanupMutex.Lock()
+
+	assignment, found := rm.assignments[getPodKey(pod)]
+	if !found {
+		rm.cleanupMutex.Unlock()
+		klog.Error("couldn't find allocation info from assignments:", getPodKey(pod))
+
+		return nil, &dpapi.UseDefaultMethodError{}
+	}
+
+	containerIndex := podCandidate.allocatedContainers
+
+	affinityMask := assignment.containers[containerIndex].tileEnv
+	getPrefDevices := assignment.containers[containerIndex].deviceIds
+
+	rm.cleanupMutex.Unlock()
+
+	devIds := request.ContainerRequests[0].DevicesIDs
+
+	// Check if all the preferred devices were also used
+	if len(devIds) != len(getPrefDevices) {
+		klog.Warningf("Allocate called with odd number of device IDs: %d vs %d", len(devIds), len(getPrefDevices))
+	}
+
+	for _, devID := range devIds {
+		if _, found := getPrefDevices[devID]; !found {
+			klog.Warningf("Not preferred device used in Allocate: %s (%v)", devID, getPrefDevices)
+		}
+	}
+
+	klog.V(4).Info("Allocate affinity mask: ", affinityMask)
+	klog.V(4).Info("Allocate device ids: ", devIds)
+
+	return rm.createAllocateResponse(devIds, affinityMask)
+}
+
+func (rm *resourceManager) GetPreferredFractionalAllocation(request *pluginapi.PreferredAllocationRequest) (
+	*pluginapi.PreferredAllocationResponse, error) {
+	response := pluginapi.PreferredAllocationResponse{}
+
+	if !isPreferredAllocationRequestOk(request, rm.skipID) {
+		// it is better to leave allocated gpu devices as is and return
+		return &response, nil
```

Comment:
Should we warn a user about the request being not ok?

Reply:
There is a warning inside "isPreferredAllocationRequestOk".